product-mini/platforms/posix/main.c

```diff
@@ -549,6 +549,18 @@ main(int argc, char *argv[])
     else
         app_instance_main(wasm_module_inst);
 
+#if WASM_ENABLE_THREAD_MGR != 0
+    /* Wait (detached) thread exit to print exception correctly.  */
+
+    wasm_module_env = wasm_runtime_get_exec_env_singleton(wasm_module_inst);
+    while (wasm_cluster_get_exec_env_count(wasm_module_env) > 1) {
+        sched_yield();
+    }
```

Comment:
sched_yield() is a common syscall (from POSIX) , but not provided on all platforms,

Reply:
OK, in fact we want to join all other threads, can you try whether this way works: add APIs in thread_manager.h/c:
```C
static void
join_thread_visitor(void *node, void *user_data)
{
    WASMExecEnv *curr_exec_env = (WASMExecEnv *)node;
    WASMExecEnv *exec_env = (WASMExecEnv *)user_data;

    if (curr_exec_env == exec_env)
        return;

    wasm_cluster_join_thread(curr_exec_env, NULL);
}

void
wasm_cluster_join_all(WASMCluster *cluster)
{
    traverse_list(&cluster->exec_env_list, join_thread_visitor, NULL);
}

void
wasm_cluster_join_all_except_self(WASMCluster *cluster, WASMExecEnv *exec_env)
{
    traverse_list(&cluster->exec_env_list, join_thread_visitor, (void *)exec_env);
}
```
And here we call `wasm_cluster_join_all_except_self(wasm_exec_env_get_cluster(exec_env), exec_env)`.

Reply:
We can't call join since in some case, the thread is detached.

Reply:
My personal understanding is that waiting for all threads to terminate is the functionality of the thread manager, we should change this component only, shouldn't change codes other places.

In wasm_cluster_join_thread, it will increase exec_env->wait_count firstly, and in wasm_cluster_detach_thread, it detects whether exec_env->wait_count is 0, if not, it won't detach the thread. Could you please try testing it? Thanks.

Reply:
> In wasm_cluster_join_thread, it will increase exec_env->wait_count firstly, and in wasm_cluster_detach_thread, it detects whether exec_env->wait_count is 0, if not, it won't detach the thread. Could you please try testing it? 

In this case we call detach first and call wasm_cluster_join_thread later, so it can't works actually.

Reply:
If wasm_cluster_detach_thread is called firstly, we may try setting the thread status, e.g. (THREAD_DETACHED) and check the status in wasm_cluster_join_thread.

Reply:
@no1wudi I discussed with @xujuntwt95329, we suggested to add the new API `wasm_cluster_wait_for_all_except_self`, and call it in wasm_application_execute_main, but not directly modify posix/main.c. We drafted the code with PR #1137, could you please help review and test whether it works?